online-norm/pytorch/online_norm_pytorch/csrc/norm/norm_cuda.cu

```diff
@@ -0,0 +1,247 @@
+    at::Tensor s_var,
+    const float afwd,
+    const float eps) {
+  CHECK_INPUT(input);
+  CHECK_INPUT(s_mu);
+  CHECK_INPUT(s_var);
+
+  const auto input_shape = input.sizes();
+  const auto inputs = input.reshape({input_shape[0], input_shape[1], -1}).toType(at::ScalarType::Float);
+
+  const auto mu = inputs.mean({2});
+  const auto var = (inputs - mu.unsqueeze(2)).pow(2).mean({2});
+
+  auto scale = at::zeros_like(mu).toType(input.scalar_type());
+  auto mean = at::zeros_like(mu).toType(input.scalar_type());
+
+  const unsigned int N = input_shape[0];
+  const unsigned int C = input_shape[1];
+  
+  const unsigned int threads = min(int(C), 1024);
+  const dim3 blocks(ceil(float(C) / threads));
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.scalar_type(), "norm_fwd", ([&] {
+    norm_fwd_kernel<scalar_t><<<blocks, threads>>>(
+        mu.data<float>(),
+        var.data<float>(),
+        s_mu.data<float>(),
+        s_var.data<float>(),
+        mean.data<scalar_t>(),
+        scale.data<scalar_t>(),
+        C, N, afwd, eps);
+  }));
+  THCudaCheck(cudaGetLastError());
+
+  const auto out = ((inputs.toType(input.scalar_type()) - mean.unsqueeze(2)) / scale.unsqueeze(2)).reshape(input_shape);
+
+  return {out, scale, s_mu, s_var};
+}
+
+template <typename scalar_t>
+__global__ void norm_uctrl_kernel(
+    const float* __restrict__ mu_delta,
+    float* __restrict__ s_u,
+    scalar_t* __restrict__ d_u,
+    const unsigned int C, const unsigned int N,
+    const float abkw) {
+
+  const int c = blockIdx.x * blockDim.x + threadIdx.x;
+  if (c < C) {
+    int idx;
+    float delta;
+    float s_u_local = s_u[c];
+
+    for(int n = 0; n < N; ++n){
+      idx = Idx2(n, c, N, C);
+      d_u[idx] = (scalar_t)(s_u_local);
+
+      delta = mu_delta[idx] - s_u_local;
+
+      s_u_local = s_u_local + (1. - abkw) * delta;
+    };
+
+    s_u[c] = s_u_local;
+  };
+}
+
+__device__ void warp_reduce(volatile float *s_mem, const unsigned int t_id, const unsigned int d) {
+  for (unsigned int ridx = 32; ridx > 0; ridx /= 2) {
+    if (d > ridx) { if (t_id < ridx) { if ((t_id + ridx) < d) { s_mem[t_id] += s_mem[t_id + ridx]; } } }
+  }
+}
+
+template <typename scalar_t>
+__global__ void norm_vctrl_kernel(
+    const scalar_t* __restrict__ grad_out,
+    float* s_v,
+    const scalar_t* __restrict__ out,
+    scalar_t* __restrict__ grad_tmp,
+    const unsigned int C, const unsigned int N, const unsigned int D,
+    const float abkw) {
+
+  const unsigned int t_id = threadIdx.x;
+  const unsigned int c = blockIdx.x;
+  const unsigned int d = blockDim.x;
+  unsigned int idx3, idx;
+
+  extern __shared__ float s_mem[];
+
+  for(int n = 0; n < N; ++n){
+    s_mem[t_id] = 0;                                // reset shared mem
+    for (idx = t_id; idx < D; idx += d) {
+      idx3 = Idx3(n, c, idx, N, C, D);              // idx in global mem
+      // vctrl logic
+      grad_tmp[idx3] = grad_out[idx3] - (1. - abkw) * (scalar_t)(s_v[c]) * out[idx3];
```

Comment:
I'm a little confused about the datatypes. I'm assuming scalar_t can be __half or float. Does the presence of abkw cast s_v[c] back to float?

Is the reason you want to support __half inside the boundaries of the layer at all performance / bandwidth saving or to try to match some other reference? It seems to require a bunch of casting.

Reply:
you're right `abkw` should be cast to `scalar_t`